src/pf/optimize.lua

```diff
@@ -176,15 +176,24 @@ local function simplify(expr, is_tail)
       if type(lhs) == 'number' and type(rhs) == 'number' then
          return assert(folders[op])(lhs, rhs)
       elseif associative_binops[op] then
-         if type(rhs) == 'table' and rhs[1] == op and type(lhs) == 'number' then
+         -- Try to make the right operand a number.
+         if type(lhs) == 'number' then
             lhs, rhs = rhs, lhs
          end
-         if type(lhs) == 'table' and lhs[1] == op and type(rhs) == 'number' then
-            if type(lhs[2]) == 'number' then
-               return { op, assert(folders[op])(lhs[2], rhs), lhs[3] }
-            elseif type(lhs[3]) == 'number' then
+         if type(lhs) == 'table' and lhs[1] == op and type(lhs[3]) == 'number' then
+            if type(rhs) == 'number' then
+               -- (A op N1) op N2 -> A op (N1 op N2)
```

Comment:
I don't think that fixpoint is required with the algorithm I have. It ensures that if there are any numbers in the tree, in the end there will be only one number as the right leaf. So there are three type of operands: bare numbers, trees without numbers, trees with one number as the right leaf. After lines 176-182 if an operand is a number, its the right one. Conditions on lines 183-193 handle the case when the left operand is a tree with one number, 194-196 - when the right operand is.

Just to check, I ran simplify once on an example of `((A op N1) op (B op (C op N2)))` (`{"+", {"+", "a", 1}, {"+", "b", {"+", "c", 2}}}`) and got `((A op (B op C)) op (N1 op N2))` (`{ "+", { "+", "a", { "+", "b", "c" } }, 3 }`) as expected.